### PR TITLE
Fix various issues when removing addons and improve state management in `addonHandler`

### LINKED ISSUE #12629: An opened file prevents add-on removal in the future
### Steps to reproduce:

Example with the Character Information add-on
1. Open a file of the add-on in an editor, e.g. `%appdata%\nvda\addons\charInfo\globalPlugins\charinfo\__init__.py`
2. Open Add-on manager and remove Character Information add-on
   The status of the add-on is "Enabled, Removed after restart"
3. Close add-on manager and restart NVDA as asked.
4. Open add-on manager
   The status of the add-on is "Enabled, Removed after restart". This is expected because the add-on could not be removed due to the file opened in the editor.
5. Close the file of the add-on opened in the editor.
6. Close the add-on manager and restart NVDA as asked.
7. Open add-on manager and check the add-on's status
8. Execute steps 5. to 7. as many times as you want.

### Actual behavior:
At step 7., the status of the add-on is "Enabled, Removed after restart".
The only way I have found to remove this add-on is to remove manually its folder or to reinstall it before removing it again.

### Expected behavior:
At step 7., the add-on should not appear anymore in the list since nothing prevents NVDA anymore to remove the add-on after step 6.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2021.1rc2
#### Windows version:
Windows 10 1809 (64-bit) build 17763.1935

#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, this error has always been present, surely in all versions of NVDA 2019.3 and beyond.

#### If add-ons are disabled, is your problem still occurring?
N/A
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by lukaszgo1**
### Link to issue number:
Fixes #12629

### Summary of the issue:
As described in #12629 it is impossible to update add-on when its folder is opened when initiating upgrade. During investigation I also encountered various other errors during add-on removal in particular:
- If disabled add-on is uninstalled and then reinstalled it stays disabled even though source code comments clearly states that it should not be happening.
- If folder of the add-on is renamed and its name no longer agrees with the name in the manifest it cannot be uninstalled via add-ons manager and there is absolutely no indication in the log as to what went wrong.

### Description of how this pull request fixes the issue:
Point 1 i.e. add-on installation cannot be finished when add-on folder was opened during installation was caused by the fact that after attempting to install add-ons pending installation list of all add-ons pending install was emptied even if the given add-on failed to install. This has been fixed by removing add-ons from list of pending installs only if installation was successful - otherwise NVDA would retry on the next restart. In addition for configurations which can be potentially affected by this I've also made sure that for all folders which name ends with suffix denoting add-ons pending installation NVDA tries to install them regardless of their presence on the list.
Point 2 disabled add-ons are not removed from the list of disabled add-ons after uninstallation was caused by the fact that there were two lists of disabled add-ons. The first one in the state and the second one `_disabledAddons` was duplicating content of the list in the state Unfortunately the second list was populated after add-ons were removed yet names of the disabled plugins were removed  from it  rather than from the list in the state. This has been fixed  by no longer using `_disabledAddons` at all - it ended up to create more confusion than it is worth and added no real benefits. I've also make sure that for old configs all add-ons which are no longer installed which are present in the list of disabled add-ons are removed.
Point 3 add-ons for which directory is named differently than the name in the manifest was solved by uninstalling add-ons after getting manifest data i.e. no longer rely on directory name matching the manifest name.
In the process I've also made addons state a class which has methods for loading, saving etc. rather than just having a simple dict  with various top level function operating on it.  This is backwards compatible.
### Testing strategy:
- Made sure that state created by previous version of NVDA can be successffully loaded
- Made sure that state created by the code from this PR can be successfully read by previous versions of NVDA
- Renamed folder of an add-on so that name of the directory is different than the name in the manifest - made sure that it can be uninstalled from add-ons manager
- Uninstalled disabled add-on - made sure that it is removed from the list of disabled add-ons in the state
- Opened folder of an add-on in Windows Explorer, tried to update the plugin, closed the opened folder - made sure that after nVDA restart add-on has been successfully installed.

### Known issues with pull request:
None known
### Change log entries:
Bug fixes
- Some bugs related to add-ons removal has been fixed.

### Code Review Checklist:

- [X] Pull Request description is up to date.
- [X] Unit tests.
- [X] System (end to end) tests.
- [X] Manual testing.
- [X] User Documentation.
- [X] Change log entry.
- [X] Context sensitive help for GUI changes.
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by lukaszgo1** [comment]
cc @seanbudd  Given you've added #12629 to 2021.3 you may want to review this. Also @CyrilleB79  your testing and review (if you have time) would be greatly appreciated.

File: source/addonHandler/__init__.py [hunk 1]
```diff
@@ -508,7 +538,6 @@ def getCodeAddon(obj=None, frameDist=1):
 	@return: L{Addon} instance or None if no code does not belong to a add-on package.
 	@rtype: C{Addon}
 	"""
-	global _availableAddons
```
Comment (lukaszgo1) [inline source/addonHandler/__init__.py:511]: This is not really related to the main problem this PR solves, but there is no reason for `_availableAddons` being global in that scope, so I've removed this line. Submitting a separate PR for a small change like this is excessive IMO.

**Comment by CyrilleB79** [comment]
I have tested the STR in https://github.com/nvaccess/nvda/issues/12629#issuecomment-904742613 on a portable copy of nvda_snapshot_pr12792-23663,da8779bd.exe. Here is my feedback:

1. I confirm that the old add-on's folder is removed and that the new add-on's installation is completed as soon as you restart without an opened folder in Windows Explorer.

2. At first restart (with Windows Explorer still opened), the following error message can be seen in nvda-old.log:
```
ERROR - core.restart (12:52:00.892) - MainThread (7012):
NVDA already in process of exiting, this indicates a logic error.
```
This needs to be clarified.

3. After first restart (with Explorer still opened), the add-on is not present at all in the add-on manager. I would have expected it to be present with a status of pending install

4. At first restart (with Windows Explorer still opened), there are many error messages indicating permission denied to copy new files. I do not know if there is a way to clarify it in the log. I am even wondering if it would be worth warning the user with a message box, e.g. "The installation of one or more add-on could not be finalized and thus are not available. Please check that you have closed all files and folders and restart NVDA".

**Comment by lukaszgo1** [comment]
> I have tested the STR in [#12629 (comment)](https://github.com/nvaccess/nvda/issues/12629#issuecomment-904742613) on a portable copy of nvda_snapshot_pr12792-23663,da8779bd.exe. Here is my feedback:

thanks for testing.

> 
>     1. I confirm that the old add-on's folder is removed and that the new add-on's installation is completed as soon as you restart without an opened folder in Windows Explorer.

Great!

> 
>     2. At first restart (with Windows Explorer still opened), the following error message can be seen in nvda-old.log:
> 
> 
> ```
> ERROR - core.restart (12:52:00.892) - MainThread (7012):
> NVDA already in process of exiting, this indicates a logic error.
> ```
> 
> This needs to be clarified.
> 

I cannot reproduce this - are you sure this was not just a coincidence?

>     1. After first restart (with Explorer still opened), the add-on is not present at all in the add-on manager. I would have expected it to be present with a status of pending install
> 

That is quite tricky to do - I'll see what I can do.

>     2. At first restart (with Windows Explorer still opened), there are many error messages indicating permission denied to copy new files. I do not know if there is a way to clarify it in the log. I am even wondering if it would be worth warning the user with a message box, e.g. "The installation of one or more add-on could not be finalized and thus are not available. Please check that you have closed all files and folders and restart NVDA".

I agree that would be nice and worth implementing at some point however there was no such warning previously. Should lack of the warning block this PR in your opinion?

**Comment by CyrilleB79** [comment]
> > 2. At first restart (with Windows Explorer still opened), the following error message can be seen in nvda-old.log:
> > ```
> > ERROR - core.restart (12:52:00.892) - MainThread (7012):
> > NVDA already in process of exiting, this indicates a logic error.
> > ```
> > This needs to be clarified.
> 
> I cannot reproduce this - are you sure this was not just a coincidence?

I cannot reproduce this again. Let's ignore it.




> > 4. At first restart (with Windows Explorer still opened), there are many error messages indicating permission denied to copy new files. I do not know if there is a way to clarify it in the log. I am even wondering if it would be worth warning the user with a message box, e.g. "The installation of one or more add-on could not be finalized and thus are not available. Please check that you have closed all files and folders and restart NVDA".
> 
> I agree that would be nice and worth implementing at some point however there was no such warning previously. Should lack of the warning block this PR in your opinion?

This PR is already an improvement (bugfix) with respect to the current situation. The warning message would be a plus and may be thought and integrated with this PR if possible. But in case of UX discussion, design  problem or whatever, it should not block this PR in my opinion.

**Comment by lukaszgo1** [comment]
@CyrilleB79 I believe I've succeeded in making add-on show up in add-ons manager with the right status i.e. "Install, enabled after restart". Can you retest?
As to the showing a warning when add-ons install fail I'm not sure about right UX here. What you've proposed i.e. "The installation of one or more add-on could not be finalized and thus are not available. Please check that you have closed all files and folders and restart NVDA"
is not really indicative as to what files / folders should be closed and getting the exact list of what is being used is quite involved. TBH I'd rather leave this as is and if you feel strongly about showing a warning please create a new issue and someone can work on this subsequently.
The amount of errors in the log is partial indication of how complex this would be - the first set is about inability to remove folder of the previous add-on version and the second group about not being able to rename folder of add-on pending install to the new name since it exists already. Showing what exactly should be closed by the user in this case is going to be difficult.

**Comment by CyrilleB79** [comment]
I have just tested the last build (portable version of nvda_snapshot_pr12792-23664,b05d9cae.exe).
The indication in the add-on manager dialog is now correct.
And as already previously, the add-on actually finishes its installation on next restart provided that no folder of the add-on is opened anymore.
Good job!

Regarding the warning, again, I think it would be better to have it but it should not block this PR which is already a progress with respect to what happens now. Just to say that IMO there is no need to specify which file or folder cannot be renamed/deleted. Someone wanting to know it may look at the log.

**Comment by lukaszgo1** [comment]
> Regarding the warning, again, I think it would be better to have it but it should not block this PR which is already a progress with respect to what happens now. Just to say that IMO there is no need to specify which file or folder cannot be renamed/deleted. Someone wanting to know it may look at the log.

You're looking at this from a  perspective of a power user for whom looking at and understanding the log is not a problem. For a power user however, this dialog is not really necessary since they would look into the log anyway so the main group for whom a graphical warning is an improvement are beginner to intermediate users.

**Comment by CyrilleB79** [comment]
If I should consider myself a power user, I would say that the friendly reminder with the warning message box is a plus, even if I can cop without it.

For someone who is not used to dig into the log, there are two possibilities:
1. I expect that such people will not encounter the issue described in #12629 because I guess they do not open the %appdata%\nvda\addons\myAddon folder at all. Or do I miss a use case?
2. If it ever happens for them, they just need to know to close all open files and folders and restart NVDA (or the computer to be sure that all is closed). Maybe the message should be refined this way.

File: source/addonHandler/__init__.py [hunk 2]
```diff
@@ -40,42 +37,77 @@
...
+		installedAddonNames = tuple(a.name for a in getAvailableAddons())
+		for disabledAddonName in list(self["disabledAddons"]):
+			if disabledAddonName not in installedAddonNames:
+				self["disabledAddons"].discard(disabledAddonName)
+
+
+state = AddonsState()
+
 
 def saveState():
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:108]: I think we should mark this for deprecation if `state.save` should replace this.

File: source/addonHandler/__init__.py [hunk 3]
```diff
@@ -40,42 +37,77 @@
...
+		during uninstallation. As a result after reinstalling add-on with the same name it was disabled
+		by default confusing users. Fix this by removing all add-ons no longer present in the config
+		from the list of disabled add-ons in the state."""
+		installedAddonNames = tuple(a.name for a in getAvailableAddons())
+		for disabledAddonName in list(self["disabledAddons"]):
+			if disabledAddonName not in installedAddonNames:
+				self["disabledAddons"].discard(disabledAddonName)
+
+
+state = AddonsState()
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:105]: I see that initialize resets the state, but the state shouldn't exist if `addonHandler` has not been initialized. I think this should be moved to initialize, and at module level this should be: 
```suggestion
state: Optional[AddonsState] = None
```

File: source/addonHandler/__init__.py [hunk 4]
```diff
@@ -40,42 +37,77 @@
...
+	def load(self) -> None:
+		"""Populates state with the default content and then loads values from the config."""
+		self.update(self._DEFAULT_STATE_CONTENT)
+		try:
+			# #9038: Python 3 requires binary format when working with pickles.
+			with open(self.statePath, "rb") as f:
+				state = pickle.load(f)
+				self.update(state)
+		except (IOError, pickle.UnpicklingError):
+			pass
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:68]: Should we log an error here?

File: source/addonHandler/__init__.py [hunk 5]
```diff
@@ -40,42 +37,77 @@
...
+
+class AddonsState(collections.UserDict):
+
+	_DEFAULT_STATE_CONTENT = {
+		"pendingRemovesSet": set(),
+		"pendingInstallsSet": set(),
+		"disabledAddons": set(),
+		"pendingEnableSet": set(),
+		"pendingDisableSet": set(),
+	}
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:52]: `SpeechState` in [speech.py](https://github.com/nvaccess/nvda/blob/master/source/speech/speech.py#L77), uses a dataclass. That way we have typed information on the items in the state that the rest of NVDA can use.

If there isn't a reason to specifically subclass `collections.UserDict`, it could be possible to just extend the dataclass with these functions, or replace the `UserDict` usage with a `self.state` variable. To convert from pickled dictionaries, dataclasses can be [converted to](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict), and [updated from](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) dictionaries.

File: source/addonHandler/__init__.py [hunk 6]
```diff
@@ -151,13 +151,11 @@ def initialize():
 	if config.isAppX:
 		log.info("Add-ons not supported when running as a Windows Store application")
 		return
-	loadState()
-	removeFailedDeletions()
-	completePendingAddonRemoves()
-	completePendingAddonInstalls()
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:157]: If we are concerned about breaking API changes here, this PR is blocked because these functions have been removed. I flag this as a concern as I'm not sure how addons like addonUpdater work.

File: source/addonHandler/__init__.py [hunk 7]
```diff
@@ -40,42 +37,77 @@
...
+					# to be a standard `dict`.
+					pickle.dump(self.data, f, protocol=0)
+			except (IOError, pickle.PicklingError):
+				log.debugWarning("Error saving state", exc_info=True)
+		else:
+			# Empty state - just delete state file and don't save anything.
+			self.removeStateFile()
+
+	def cleanupRemovedDisabledAddons(self) -> None:
+		"""Versions of NVDA before 2021.3 failed to remove add-on from list of disabled add-ons
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:95]: I think it's more accurate to reference the PR here.

```suggestion
		"""Versions of NVDA before #12792 failed to remove add-on from list of disabled add-ons
```

File: source/addonHandler/__init__.py (hunk 5 above)
Comment (seanbudd) [reply source/addonHandler/__init__.py:52]: A reason to keep using `collections.UserDict` is to maintain backward compatibility with `state`. I think this should be documented under the class declaration if that is the case.

File: source/addonHandler/__init__.py [hunk 8]
```diff
@@ -210,23 +229,27 @@ def _getAvailableAddonsFromPath(path):
 					yield a
 				except:
 					log.error("Error loading Addon from path: %s", addon_path, exc_info=True)
-				
+
 _availableAddons = collections.OrderedDict()
-def getAvailableAddons(refresh=False, filterFunc=None):
+
+
+def getAvailableAddons(refresh=False, filterFunc=None, isFirstLoad=False):
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:236]: Could the parameter and return types be updated here - at least for the new variable.

**Review by seanbudd — COMMENTED** [review]
I have a few concerns around backwards compatibility, which I have raised below. If it can't be maintained easily here, this can be looked at as a priority for 2022.1 - which isn't too far away from starting.

**Comment by seanbudd** [comment]
I also agree that a warning message may be helpful here, but it won't block this getting merged. A new issue should be raised for it, and perhaps a code comment on where it should go.

File: source/addonHandler/__init__.py (hunk 3 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:105]: This has been done to  preserve backwards compatibility. Currently on master state is initialized to an empty `dict` at the module level.

File: source/addonHandler/__init__.py (hunk 4 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:68]: I don't think so. If we cannot unpickle the state or the state file does not exist empty state is created and error which occurred is not really actionable.

File: source/addonHandler/__init__.py (hunk 5 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:52]: Is preserving backward compatibility enough of a reason to keep subclassing `collections.UserDict` in your opinion?

File: source/addonHandler/__init__.py (hunk 6 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:157]: Addon Updater is not using any of these. In fact I don't think any  half serious add-on would since they modify content of addons folder and it is not safe to call them after add-ons were loaded. I don't think we should be voried about backwards compatibility for these removals.

File: source/addonHandler/__init__.py (hunk 2 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py]: Done as of 2dd5b7e7a

File: source/addonHandler/__init__.py (hunk 8 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py]: Done as of cae49fbd2

File: source/addonHandler/__init__.py (hunk 5 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:56]: I've added an explanatory comment.

**Comment by lukaszgo1** [comment]
@seanbudd I believe all your review comments are now addressed.

File: source/addonHandler/__init__.py (hunk 4 above)
Comment (seanbudd) [reply source/addonHandler/__init__.py:72]: Yes but if a user encounters this problem it might be helpful when debugging to know that is occurring. For example, if we encounter a bug with the state being written to incorrectly, and so the state keeps being overwritten with an empty state.
I would log each error separately, the IO error at a debug level and the unpickling issue at a warning level.

File: source/addonHandler/__init__.py [hunk 9]
```diff
@@ -40,42 +40,80 @@
...
-	statePath=os.path.join(globalVars.appArgs.configPath,stateFilename)
-	try:
-		# #9038: Python 3 requires binary format when working with pickles.
-		with open(statePath, "wb") as f:
-			pickle.dump(state, f, protocol=0)
-	except:
-		log.debugWarning("Error saving state", exc_info=True)
+
+class AddonsState(collections.UserDict):
+	"""Subclasses `collections.UserDict` to preserver backwards compatibility."""
```
Comment (seanbudd) [inline source/addonHandler/__init__.py:48]:
```suggestion
	"""Subclasses `collections.UserDict` to preserve backwards compatibility."""
```

File: source/addonHandler/__init__.py (hunk 6 above)
Comment (seanbudd) [reply source/addonHandler/__init__.py:157]: I agree that this is the case for `removeFailedDeletions`,  `completePendingAddonRemoves` and `completePendingAddonInstalls` but what about `loadState`? Should this get the same treatment as `saveState`? It's unlikely that either of these are used but they should be safe to call as far as I can tell.

File: source/addonHandler/__init__.py (hunk 3 above)
Comment (seanbudd) [reply source/addonHandler/__init__.py:109]: This module should be initialized before any addons are loaded so I think this should be safe, but it's a minor issue and perhaps not worth the small risk.

File: source/addonHandler/__init__.py (hunk 5 above)
Comment (seanbudd) [reply source/addonHandler/__init__.py:56]: Yes, it's just helpful to document why the usage is different to other parts of the codebase. I also wasn't sure if the intention was to change the API.

File: source/addonHandler/__init__.py (hunk 3 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:109]: I'd rather leave it as is just to be on the safe side.

File: source/addonHandler/__init__.py (hunk 4 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py]: Done as of 2cf763f61

File: source/addonHandler/__init__.py (hunk 6 above)
Comment (lukaszgo1) [reply source/addonHandler/__init__.py:157]: Done as of 38060bc00. I'm not convincet this has been necesary but better to be safe than sorry.

**Review by seanbudd — APPROVED** [review]

**Comment by lukaszgo1** [comment]
> Changes
> 
> ```
> - If a disabled addon is uninstalled and then re-installed it is re-enabled. (#12792)
> ```
> 

IMO this should be a bug fix - this was always the intent it just never worked as it should.

> Bug fixes
> 
> ```
> - Fixed bugs around updating and removing addons where the addon folder has been renamed or has files opened. (#12792, #12629)
> ```
> 
> Changes for developers
> 
> ```
> - The usage of functions ``addonHandler.loadState`` and ``addonHandler.saveState`` should be replaced with their equivalents ``addonHandler.state.save`` and ``addonHandler.state.load`` before 2022.1. (#12792)
> ```


The rest looks pretty nice.
